src/bun.js/node/types.zig

```diff
@@ -1835,548 +1849,2876 @@ pub const Emitter = struct {
+
+            // Translated from the following JS code:
+            //   resolvedPath = `${path}/${resolvedPath}`;
+            const resolvedPathLen = resolvedPath.len;
+            if (resolvedPathLen > 0) {
+                // Move all bytes to the right by path.len + 1
+                bufOffset = len + 1;
+                var j: usize = resolvedPathLen - 1;
+                while (j >= 0) : (j -= if (j > 0) 1 else break) {
+                    tmpBuf3[j + bufOffset] = resolvedPath[j];
+                }
+            }
+            bufSize = len;
+            @memcpy(tmpBuf3[0 ..bufSize], path);
+            bufSize += 1;
+            tmpBuf3[len] = CHAR_FORWARD_SLASH;
+            bufSize += resolvedPathLen;
+
+            resolvedPath = tmpBuf3[0 ..bufSize];
+            resolvedAbsolute = path[0] == CHAR_FORWARD_SLASH;
+        }
+
+        // Exit early for empty path.
+        if (resolvedPath.len == 0) {
+            return CHAR_STR_DOT;
+        }
+
+        // At this point the path should be resolved to a full absolute path, but
+        // handle relative paths to be safe (might happen when process.cwd() fails)
+
+        // Normalize the path
+        const bufBackedNormalizedPath = @This().normalizeStringImpl(T, resolvedPath, !resolvedAbsolute, CHAR_FORWARD_SLASH, .posix, buf);
+        // Move resolvedPath from bufBackedNormalizedPath to tmpBuf3
+        bufSize = bufBackedNormalizedPath.len;
+        @memcpy(tmpBuf3[0 ..bufSize], bufBackedNormalizedPath);
+        resolvedPath = tmpBuf3[0 ..bufSize];
+
+        // Translated from the following JS code:
+        //   if (resolvedAbsolute) {
+        //     return `/${resolvedPath}`;
+        //   }
+        if (resolvedAbsolute) {
+            bufSize = 1;
+            buf[0] = CHAR_FORWARD_SLASH;
+            bufOffset = bufSize;
+            bufSize += resolvedPath.len;
+            @memcpy(buf[bufOffset ..bufSize], resolvedPath);
+            return buf[0 ..bufSize];
+        }
+        // Translated from the following JS code:
+        //   return resolvedPath.length > 0 ? resolvedPath : '.';
+        return if (resolvedPath.len > 0) resolvedPath else CHAR_STR_DOT;
+    }
+
+    /// Based on Node v21.6.1 path.win32.resolve:
+    /// https://github.com/nodejs/node/blob/6ae20aa63de78294b18d5015481485b7cd8fbb60/lib/path.js#L162
+    pub fn resolveWindowsImpl(comptime T: type, paths: []const []const T, buf: []T) []const T {
+        const isSepImpl = @This().isSepWindowsImpl;
+
+        // Backed by smallTmpBuf2
+        var resolvedDevice: []const T = "";
+        // Backed by expandable tmpBuf3 because resolvedTail may be long.
+        // We use tmpBuf3 here because resolvePosixImpl is called by other methods and using
+        // tmpBuf3 here avoids stepping on others' toes.
+        var resolvedTail: []const T = "";
+        var resolvedAbsolute: bool = false;
+
+        var bufOffset: usize = 0;
+        var bufSize: usize = 0;
+
+        var envPath: ?[]const T = null;
```

Comment:
envPath is... interesting. i really wish we didn't have to do this check. it is most certainly the most expensive part of this function, querying the environment. and path.resolve is done EVERYWHERE, so it needs to be fast.

node.js is lucky because process.env is automatically cached and is already a nice hash table for them. we dont have that, the call to getEnvVarOwned is quite something.

i kind of want to say "what if we TODO comment this for now" and omit the functionality. (no one has asked for it)

if we really really want this (im against having it for now), then the arena allocator is silly. we can just use a regular allocator / stack fallback of something really small like 512 bytes (this var is going to be short). cleanup logic is alright but i think the  entire allocator can be moved right into the block where envPath is init, and we deinit there (the data is copied out by the end of that inner block, no?). Then, im not 1000% sure of this but i think all this logic doesnt need to happen if the resolved path is already at some point absolute, so this work can be filtered away even more.

Reply:
We could use Zig's EnvMap which is a snapshot of the Env vars:
https://github.com/ziglang/zig/blob/f2e249e92020462ac29f34fc3d45cef3dc90a3b2/lib/std/process.zig#L249-L253

This is also only in the Win32 branch. The logic is how Node's implementation has had it for over 9 years or so and was a perf improvement over what they had before that:
https://github.com/nodejs/node/blame/5cb5b1f8c4b67d31d977cee66cb07adaa4acb25f/lib/path.js#L174

I'm super into using the right alloc for the occasion and looking for all the guidance in that area 🙇‍♂️

Reply:
I optimized the windows branch and made the other platforms skip it.